connmgr/connmanager_test.go

```diff
@@ -0,0 +1,373 @@
+
+package connmgr_test
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"net"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/btcsuite/btcd/connmgr"
+	"github.com/btcsuite/btclog"
+)
+
+const maxRetryDuration = 2 * time.Millisecond
+
+func init() {
+	connmgr.TstSetMaxRetryDuration(maxRetryDuration)
+}
+
+// mockAddr mocks a network address
+type mockAddr struct {
+	net, address string
+}
+
+func (m mockAddr) Network() string { return m.net }
+func (m mockAddr) String() string  { return m.address }
+
+// mockConn mocks a network connection by implementing the net.Conn interface.
+type mockConn struct {
+	io.Reader
+	io.Writer
+	io.Closer
+
+	// local network, address for the connection.
+	lnet, laddr string
+
+	// remote network, address for the connection.
+	rnet, raddr string
+}
+
+// LocalAddr returns the local address for the connection.
+func (c mockConn) LocalAddr() net.Addr {
+	return &mockAddr{c.lnet, c.laddr}
+}
+
+// RemoteAddr returns the remote address for the connection.
+func (c mockConn) RemoteAddr() net.Addr {
+	return &mockAddr{c.rnet, c.raddr}
+}
+
+// Close handles closing the connection.
+func (c mockConn) Close() error {
+	return nil
+}
+
+func (c mockConn) SetDeadline(t time.Time) error      { return nil }
+func (c mockConn) SetReadDeadline(t time.Time) error  { return nil }
+func (c mockConn) SetWriteDeadline(t time.Time) error { return nil }
+
+// mockDialer mocks the net.Dial interface by returning a mock connection to
+// the given address.
+func mockDialer(network, address string) (net.Conn, error) {
+	r, w := io.Pipe()
+	c := &mockConn{raddr: address}
+	c.Reader = r
+	c.Writer = w
+	return c, nil
+}
+
+// TestNewConfig tests that new ConnManager config is validated as expected.
+func TestNewConfig(t *testing.T) {
+	_, err := connmgr.New(&connmgr.Config{})
+	if err == nil {
+		t.Fatalf("connmgr.New expected error: 'Dial can't be nil', got nil")
+	}
+	_, err = connmgr.New(&connmgr.Config{
+		Dial: mockDialer,
+	})
+	if err != nil {
+		t.Fatalf("connmgr.New unexpected error: %v", err)
+	}
+}
+
+// TestUseLogger tests that a logger can be passed to UseLogger
+func TestUseLogger(t *testing.T) {
+	l, err := btclog.NewLoggerFromWriter(bytes.NewBuffer(nil), btclog.InfoLvl)
+	if err != nil {
+		t.Fatal(err)
+	}
+	connmgr.UseLogger(l)
+}
+
+// TestStartStop tests that the connection manager starts and stops as
+// expected.
+func TestStartStop(t *testing.T) {
+	connected := make(chan *connmgr.ConnReq)
+	cmgr, err := connmgr.New(&connmgr.Config{
+		MaxOutbound:   1,
+		GetNewAddress: func() string { return "127.0.0.1:18555" },
+		Dial:          mockDialer,
+		OnConnection: func(c *connmgr.ConnReq, conn net.Conn) {
+			connected <- c
+		},
+	})
+	if err != nil {
+		t.Fatalf("connmgr.New error: %v", err)
+	}
+	cmgr.Start()
+	go cmgr.Wait()
+	gotConnReq := <-connected
+	cmgr.Stop()
+	// already stopped
+	cmgr.Stop()
+	// ignored
+	cr := &connmgr.ConnReq{Addr: "127.0.0.1:18555", Permanent: true}
+	cmgr.Connect(cr)
```

Comment:
This section doesn't actually appear to be testing the result?  If it wasn't working properly and ignoring the request due to it being stopped, how would the test know?